A68-deterministic-subsetting-lb-policy.md

```diff
@@ -0,0 +1,177 @@
+```
+
+Here is the implementation of this algorithm in pseudocode with detailed comments. 
+
+```
+func filter_addresses(addresses, subset_size, client_index, sort_addresses)
+    backend_count = addresses.length()
+    if backend_count > subset_size {
+        // if we don't have enough addresses to cover the desired subset size, just return the whole list
+        return addresses
+    }
+
+    if sort_addresses {
+        // sort address list because the algorithm assumes that the initial
+        // order of the addresses is the same for every client
+        addresses.sort()
+    }
+
+    // subset_count indicates how many clients we can have so that every client is connected to exactly 
+    // subset_size distinct backends and no 2 clients connect to the same backend.
+    subset_count = backend_count / subset_size
+    
+    // Given subset_count we now can divide clients by rounds. Every round have exactly subset_count clients.
+    // round indicates the index of the round for the current client based on its index.
+    round = client_index / subset_count
+
+    // There might be some lefover backends withing every round in cases when backend_count % subset_size != 0.
+    // excluded_count indicates how many leftover backends we have on every round.
+    excluded_count = backend_count % subset_size
+
+    // We want to choose what backends are excluded in a round robin fashion before shufling the backends.
+    // This way we make sure that for any 2 backends the difference of how many times they are excluded is at most 1.
+    // excluded_start indicates the index of the first excluded backend
+    excluded_start := (round * excluded_count) % backend_count
+    // excluded_start indicates the index after the last excluded backend.
+    // It could wrap around the end of the addresses list.
+    excluded_end := (excluded_start + excluded_count) % backend_count
+
+    if excluded_start < excluded_end {
+        // excluded_end doesn't wrap, exclude addresses from the interval [excluded_start, excluded_end)
+    } else {
+        // excluded_end wraps around the end of the addresses list, exclude intervals [0:excluded_start] and [excluded_end, end_of_the_array)
+    }
+
+    // randomly shuffle the addresses to increase subset diversity. Use round as seed to make sure
+    // that clients within the same round shuffle addresses identically. 
+    addresses.shuffle(seed: round)
+
+    // subset_id is the index for the current client withing the round
+    subset_id := client_index % subset_count
+
+    // calculate start start and end of the resulting subset 
+    start = subsetId * subset_size
+    end = start + subset_size
+    return addresses[start: end]
```

Comment:
After we select which addresses, we could map the addresses back to their original order. Is that worthwhile?